### PR TITLE
Issue 12677 role selector workflow step editor

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 1]
```diff
@@ -70,6 +136,9 @@
 
 	protected final DotConcurrentFactory concurrentFactory = DotConcurrentFactory.getInstance();
 
+	private List <Role> specialRolesHierarchy;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: set to volatile

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 2]
```diff
@@ -2100,32 +2170,125 @@ public void updateStepReferences(final String stepId, final String replacementSt
...
+
+	/**
+	 * Returns an immutable list with the Hierarchy of special roles according to the permission precedence.
+	 * This means that the position roles[i] has more precedence in that roles[i+1]
+	 * @return a list of special roles ordered by permission precedence.
+	 * @throws DotDataException
+	 */
+	private synchronized List<Role> getSpecialRolesHierarchy() throws DotDataException {
+
+		if (specialRolesHierarchy == null) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: I would use the double check algorithm

if (x == null) {
   sync {
     if (x == null) {

I would also use a java.util.Set

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 3]
```diff
@@ -2100,32 +2170,125 @@ public void updateStepReferences(final String stepId, final String replacementSt
...
+
+	/**
+	 * Given a set roles this method verifies if an action has access to at least one of them.
+	 * @param wa the workflow action to examine
+	 * @param roles the roles to apply
+	 * @return true if any of the roles is met by the action
+	 * @throws DotDataException
+	 */
+	private boolean hasAnyRolePermission(final WorkflowAction wa,final Iterable<Role> roles) throws DotDataException{
+        for(Role role:roles){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: set to final

Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: done

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 2 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: I'm Keeping the List as order matters and it only has 4 elements

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: done

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 2 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: double check also ... done

**Review by jdotcms — APPROVED** [review]

**Review by jgambarios — APPROVED** [review]

**Pull request merged**